File: Leetcode/785/785.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        constexpr int INF = 0x3f3f3f3f;
        int n = (int)graph.size();
        vector<int> color(n, INF);
        bool result = true;
        // bfs from 0
        queue<int> q;
        for (int s = 0; s < n; ++s) {
            if (color.at(s) == INF) {
                color.at(s) = 0;
                q.push(s);
            }
            while (!q.empty()) {
                int u = q.front();
                q.pop();
                for (const auto &v : graph.at(u)) {
                    if (color.at(v) == INF) {
                        color.at(v) = 1 - color.at(u);
                        q.push(v);
                    } else if (color.at(v) == color.at(u)) {
                        result = false;
                        return result;
                    }
                }
            }
        }
        return result;
    }
};

File: Leetcode/785/785.go
var (
	hasAns bool
	color  []int	// 0/1/-1
)

func dfs(adj *[][]int, u int) {
	if !hasAns {
		return
	}
	if color[u] == 1 {
		for i := 0; i < len((*adj)[u]); i++ {
			v := (*adj)[u][i]
			if color[v] == 0 {
				color[v] = -1
				dfs(adj, v)
			} else if color[v] == 1 {
				hasAns = false
				return
			}
		}
	} else {
		for i := 0; i < len((*adj)[u]); i++ {
			v := (*adj)[u][i]
			if color[v] == 0 {
				color[v] = 1
				dfs(adj, v)
			} else if color[v] == -1 {
				hasAns = false
				return
			}
		}
	}
}

func isBipartite(graph [][]int) bool {
	hasAns = true
	color = nil
	n := len(graph)
	color = make([]int, n)
	for i := 0; i < n; i++ {
		if color[i] == 0 {
			color[i] = 1
			dfs(&graph, i)
		}
	}
	return hasAns
}